File: server/src/main/java/io/microconfig/server/rest/exceptions/ExceptionsHandler.java
package io.microconfig.server.rest.exceptions;

import io.microconfig.server.configs.ComponentNotFoundException;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.bind.annotation.RestController;

@ControllerAdvice
@RestController
@Slf4j
public class ExceptionsHandler {
    private static final ResponseEntity<String> BAD_REQUEST = new ResponseEntity<>(HttpStatus.BAD_REQUEST);
    private static final ResponseEntity<String> NOT_FOUND = new ResponseEntity<>(HttpStatus.NOT_FOUND);
    private static final ResponseEntity<String> FORBIDDEN = new ResponseEntity<>(HttpStatus.FORBIDDEN);
    private static final ResponseEntity<String> TEAPOT = new ResponseEntity<>(HttpStatus.I_AM_A_TEAPOT);

    @ExceptionHandler(Exception.class)
    public ResponseEntity<String> handle(Exception ex) {
        log.trace("Exception on rest call", ex);
        return response(ex);
    }

    private ResponseEntity<String> response(Exception ex) {
        if (ex instanceof NotFoundException) {
            return NOT_FOUND;
        }

        if (ex instanceof ForbiddenException) {
            return FORBIDDEN;
        }

        if (ex instanceof ComponentNotFoundException) {
            return TEAPOT;
        }

        return BAD_REQUEST;
    }
}

File: server/src/main/kotlin/io/microconfig/server/api/exceptions/ExceptionsHandler.kt
package io.microconfig.server.api.exceptions

import io.microconfig.core.environments.repository.EnvironmentException
import io.microconfig.core.properties.ResolveException
import io.microconfig.core.properties.repository.ComponentNotFoundException
import io.microconfig.server.common.logger
import io.microconfig.server.git.exceptions.RefNotFound
import org.springframework.http.HttpStatus
import org.springframework.http.HttpStatus.*
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ControllerAdvice
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestController

@ControllerAdvice
@RestController
class ExceptionsHandler {
    val log = logger()

    @ExceptionHandler(Exception::class)
    fun handle(ex: Exception): ResponseEntity<ErrorResponse> {
        logEx(ex)
        return when (ex) {
            is ResolveException -> response(ex.message, BAD_REQUEST)
            is RefNotFound -> response(ex.message, BAD_REQUEST)
            is EnvironmentException -> response(ex.message, BAD_REQUEST)
            is ServerErrorException -> response(ex.message, INTERNAL_SERVER_ERROR)
            is NotFoundException -> response(ex.message, NOT_FOUND)
            is ComponentNotFoundException -> response(ex.message, NOT_FOUND)
            is ForbiddenException -> response(ex.message, FORBIDDEN)
            is BadRequestException -> response(ex.message, BAD_REQUEST)
            else -> response("Something went wrong", BAD_REQUEST)
        }
    }

    private fun response(error: String?, status: HttpStatus): ResponseEntity<ErrorResponse> {
        return ResponseEntity<ErrorResponse>(ErrorResponse(error), status)
    }

    private fun logEx(ex: Exception) {
        val c = ex.javaClass.name
        when {
            c.startsWith("io.microconfig") -> log.error(ex.message)
            else -> log.error("$c: ${ex.message} at ${ex.stackTrace[0]}")
        }
    }

    data class ErrorResponse(val error: String?)

}